orchagent/routeorch.cpp

```diff
@@ -2106,3 +2112,743 @@ bool RouteOrch::removeOverlayNextHops(sai_object_id_t vrf_id, const NextHopGroup
+            {
+                removeNextHopGroup(it_nhg.first);
+            }
+        }
+    }
+}
+
+void RouteOrch::addTempLabelRoute(LabelRouteBulkContext& ctx, const NextHopGroupKey &nextHops)
+{
+    SWSS_LOG_ENTER();
+
+    Label& label = ctx.label;
+
+    auto next_hop_set = nextHops.getNextHops();
+
+    /* Remove next hops that are not in m_syncdNextHops */
+    for (auto it = next_hop_set.begin(); it != next_hop_set.end();)
+    {
+        if (!m_neighOrch->hasNextHop(*it))
+        {
+            SWSS_LOG_INFO("Failed to get next hop %s for %u",
+                   (*it).to_string().c_str(), label);
+            it = next_hop_set.erase(it);
+        }
+        else
+            it++;
+    }
+
+    /* Return if next_hop_set is empty */
+    if (next_hop_set.empty())
+        return;
+
+    /* Randomly pick an address from the set */
+    auto it = next_hop_set.begin();
+    advance(it, rand() % next_hop_set.size());
+
+    /* Set the route's temporary next hop to be the randomly picked one */
+    NextHopGroupKey tmp_next_hop((*it).to_string());
+    ctx.tmp_next_hop = tmp_next_hop;
+
+    addLabelRoute(ctx, tmp_next_hop);
+}
+
+bool RouteOrch::addLabelRoute(LabelRouteBulkContext& ctx, const NextHopGroupKey &nextHops)
+{
+    SWSS_LOG_ENTER();
+
+    sai_object_id_t& vrf_id = ctx.vrf_id;
+    Label& label = ctx.label;
+
+    /* next_hop_id indicates the next hop id or next hop group id of this route */
+    sai_object_id_t next_hop_id;
+
+    if (m_syncdLabelRoutes.find(vrf_id) == m_syncdLabelRoutes.end())
+    {
+        m_syncdLabelRoutes.emplace(vrf_id, LabelRouteTable());
+        m_vrfOrch->increaseVrfRefCount(vrf_id);
+    }
+
+    auto it_route = m_syncdLabelRoutes.at(vrf_id).find(label);
+
+    /* The route is pointing to a next hop */
+    if (nextHops.getSize() == 1)
+    {
+        const NextHopKey& nexthop = *nextHops.getNextHops().begin();
+        if (nexthop.isIntfNextHop())
+        {
+            next_hop_id = m_intfsOrch->getRouterIntfsId(nexthop.alias);
+            /* rif is not created yet */
+            if (next_hop_id == SAI_NULL_OBJECT_ID)
+            {
+                SWSS_LOG_INFO("Failed to get next hop %s for %u",
+                        nextHops.to_string().c_str(), label);
+                return false;
+            }
+        }
+        else
+        {
+            if (m_neighOrch->hasNextHop(nexthop))
+            {
+                next_hop_id = m_neighOrch->getNextHopId(nexthop);
+            }
+            /* See if there is an IP neighbor nexthop */
+            else if (nexthop.isMplsNextHop() &&
+                     m_neighOrch->hasNextHop(NextHopKey(nexthop.ip_address, nexthop.alias)))
+            {
+                m_neighOrch->addNextHop(nexthop);
+                next_hop_id = m_neighOrch->getNextHopId(nexthop);
+            }
+            else
+            {
+                SWSS_LOG_INFO("Failed to get next hop %s for %u",
+                        nextHops.to_string().c_str(), label);
```

Comment:
Can we change these log info to error for all error scenarios? When NH/Neighbor entry is resolved, are we pushing this label route?

Reply:
The log level is the same as existing Route code.